File: Python/868A-barktounlock.py
passw = input()
n = int(input())

if n == 1:
    s = input()
    print('YES' if (s == passw or s == passw[::-1]) else 'NO')
else:
    flag = 0
    okf = 0
    okl = 0
    while n > 0:
        s = input()
        if passw[0] == s[1]:
            okf = 1
        if passw[1] == s[0]:
            okl = 1
        if passw == s:
            flag = 1
        n -= 1
    print("YES" if (flag == 1 or (okf == okl and okf == 1))else "NO")

File: C++/A/868A-barktounlock.cpp
#include <bits/stdc++.h>
using namespace std;
int main()
{
    int n;
    string s, pass;
    cin >> pass >> n;
    if (n == 1)
    {
        cin >> s;
        if (s == pass)
            cout << "YES";
        else
        {
            reverse(s.begin(), s.end());
            cout << ((s == pass) ? "YES" : "NO");
        }
    }
    else
    {
        int flag = 0, okf = 0, okl = 0;
        while (n--)
        {
            cin >> s;
            if (pass[0] == s[1])
                okf = 1;
            if (pass[1] == s[0])
                okl = 1;
            if (pass == s)
                flag = 1;
        }
        cout << (((flag == 1 || (okf == okl && okf == 1))) ? "YES" : "NO");
    }
}